sound/soc/sof/intel/hda-dai.c

```diff
@@ -0,0 +1,136 @@
+ *
+ * Copyright(c) 2018 Intel Corporation. All rights reserved.
+ *
+ * Authors: Keyon Jie <yang.jie@linux.intel.com>
+ */
+
+#include <sound/pcm_params.h>
+#include "../sof-priv.h"
+
+/* common dai driver for skl+ platforms */
+static const struct snd_soc_dai_ops skl_dai_ops = {
+};
+
+#define SKL_FORMATS (SNDRV_PCM_FMTBIT_S16_LE | SNDRV_PCM_FMTBIT_S24_LE | \
+	SNDRV_PCM_FMTBIT_S32_LE | SNDRV_PCM_FMTBIT_FLOAT)
+
+/*
+ * Skylake+ DAIs, some products who use this DAI array only physically
+ * have a subset of the DAIs, but no harm is done here by adding the whole
+ * set.
+ */
+static struct snd_soc_dai_driver skl_dai[] = {
+{
+	.name = "SSP0 Pin",
+	.ops = &skl_dai_ops,
+	.playback = SOF_DAI_STREAM("ssp0 Tx", 1, 8,
+				   SNDRV_PCM_RATE_8000_192000, SKL_FORMATS),
+	.capture = SOF_DAI_STREAM("ssp0 Rx", 1, 8,
+				  SNDRV_PCM_RATE_8000_192000, SKL_FORMATS),
+},
+{
+	.name = "SSP1 Pin",
+	.ops = &skl_dai_ops,
+	.playback = SOF_DAI_STREAM("ssp1 Tx", 1, 8,
+				   SNDRV_PCM_RATE_8000_192000, SKL_FORMATS),
+	.capture = SOF_DAI_STREAM("ssp1 Rx", 1, 8,
+				  SNDRV_PCM_RATE_8000_192000, SKL_FORMATS),
+},
+{
+	.name = "SSP2 Pin",
+	.ops = &skl_dai_ops,
+	.playback = SOF_DAI_STREAM("ssp2 Tx", 1, 8,
+				   SNDRV_PCM_RATE_8000_192000, SKL_FORMATS),
+	.capture = SOF_DAI_STREAM("ssp2 Rx", 1, 16,
+				  SNDRV_PCM_RATE_8000_192000, SKL_FORMATS),
+},
+{
+	.name = "SSP3 Pin",
+	.ops = &skl_dai_ops,
+	.playback = SOF_DAI_STREAM("ssp3 Tx", 1, 8,
+				   SNDRV_PCM_RATE_8000_192000, SKL_FORMATS),
+	.capture = SOF_DAI_STREAM("ssp3 Rx", 1, 8,
+				  SNDRV_PCM_RATE_8000_192000, SKL_FORMATS),
+},
+{
+	.name = "SSP4 Pin",
+	.ops = &skl_dai_ops,
+	.playback = SOF_DAI_STREAM("ssp4 Tx", 1, 8,
+				   SNDRV_PCM_RATE_8000_192000, SKL_FORMATS),
+	.capture = SOF_DAI_STREAM("ssp4 Rx", 1, 8,
+				  SNDRV_PCM_RATE_8000_192000, SKL_FORMATS),
+},
+{
+	.name = "SSP5 Pin",
+	.ops = &skl_dai_ops,
+	.playback = SOF_DAI_STREAM("ssp5 Tx", 1, 8,
+				   SNDRV_PCM_RATE_8000_192000, SKL_FORMATS),
+	.capture = SOF_DAI_STREAM("ssp5 Rx", 1, 8,
+				  SNDRV_PCM_RATE_8000_192000, SKL_FORMATS),
+},
+{
+	.name = "DMIC01 Pin",
+	.ops = &skl_dai_ops,
+	.capture = SOF_DAI_STREAM("DMIC01 Rx", 1, 4,
+				  SNDRV_PCM_RATE_8000_192000, SKL_FORMATS),
+},
+{
+	.name = "DMIC16k Pin",
+	.ops = &skl_dai_ops,
+	.capture = SOF_DAI_STREAM("DMIC16k Rx", 1, 4,
+				  SNDRV_PCM_RATE_8000_192000, SKL_FORMATS),
+},
+{
+	.name = "iDisp1 Pin",
+	.ops = &skl_dai_ops,
+	.playback = SOF_DAI_STREAM("iDisp1 Tx", 1, 8,
+				   SNDRV_PCM_RATE_8000_192000, SKL_FORMATS),
+},
+{
+	.name = "iDisp2 Pin",
+	.ops = &skl_dai_ops,
+	.playback = SOF_DAI_STREAM("iDisp2 Tx", 1, 8,
+				   SNDRV_PCM_RATE_8000_192000, SKL_FORMATS),
+},
+{
+	.name = "iDisp3 Pin",
+	.ops = &skl_dai_ops,
+	.playback = SOF_DAI_STREAM("iDisp3 Tx", 1, 8,
+				   SNDRV_PCM_RATE_8000_192000, SKL_FORMATS),
+},
+{
+	.name = "Analog Codec DAI",
+	.ops = &skl_dai_ops,
+	.playback = SOF_DAI_STREAM("Analog Codec Playback", 1, 2,
```

Comment:
I believe the max number of channels should be 16 for HDAudio (same for all digital parts below)

Reply:
please feel free to change it, as I am not able to test them yet, so I can't give any comment about it.